File: lib/devices/wearos/onboard_wearos_watch_use_case.dart
import '../../domain/model/ble_sensor_models.dart';
import '../core/pairing/watch_pairing_port.dart';
import '../core/registry/ble_device_repository.dart';

/// The one platform step WearOS onboarding shows — the companion association
/// dialog. There is no bond step (a WearOS watch has no GFDI auth boundary to
/// protect) and no transport probe (no GFDI to enumerate).
enum WearosOnboardStep { associating }

/// WearOS onboarding always succeeds once the user has picked the watch — there
/// is nothing that can fail the way a Garmin bond can. [associated] records
/// whether the optional companion association was granted.
class WearosOnboardOutcome {
  const WearosOnboardOutcome({required this.device, required this.associated});

  final BleSensorDevice device;
  final bool associated;
}

/// Turns a scanned WearOS smartwatch (Galaxy, Pixel, …) into a registered
/// `(watch, wearos)` device — the sibling of [OnboardGarminWatchUseCase] built
/// on the same [WatchPairingPort] + [BleDeviceRepository] seam.
///
/// Two steps, not the Garmin four:
///   1. **Associate.** The companion dialog — optional in every direction; a
///      false (declined, or unsupported) is recorded, never raised. It is the
///      "pair like Garmin" parity, minus the security bond Garmin needs.
///   2. **Register.** As [BleDeviceKind.watch] with [DeviceIntegration.wearos]
///      and NO capabilities, so it is off the Garmin sync path
///      ([BleSensorDevice.isGarminWatch]) and out of capability assignment.
///
/// No bond and no GFDI probe: a WearOS watch speaks neither. Its live heart rate
/// comes over standard GATT and its recorded data through Health Connect.
class OnboardWearosWatchUseCase {
  const OnboardWearosWatchUseCase(this._pairing, this._bleDeviceRepository);

  final WatchPairingPort _pairing;
  final BleDeviceRepository _bleDeviceRepository;

  Future<WearosOnboardOutcome> call(
    BleDiscoveredDevice device, {
    required String displayName,
    void Function(WearosOnboardStep)? onStep,
  }) async {
    onStep?.call(WearosOnboardStep.associating);
    var associated = false;
    try {
      associated = await _pairing.associateCompanion(device.address, displayName);
    } catch (_) {
      // The companion association is a best-effort nicety; the watch onboards
      // whether or not the platform granted (or even offered) it.
    }

    final registered = _bleDeviceRepository.addDevice(
      displayName: displayName,
      address: device.address,
      bluetoothName: device.name,
      capabilities: const {},
      kind: BleDeviceKind.watch,
      integration: DeviceIntegration.wearos,
    );
    return WearosOnboardOutcome(device: registered, associated: associated);
  }

  /// Undoes [call] at the OS level — drops the companion association. The
  /// registry entry is removed by the usual forget path.
  Future<void> forget(String address) =>
      _pairing.disassociateCompanion(address);
}

File: app/src/main/kotlin/tech/mmarca/openvitals/devices/wearos/OnboardWearOsWatchUseCase.kt
package tech.mmarca.openvitals.devices.wearos

import javax.inject.Inject
import javax.inject.Singleton
import tech.mmarca.openvitals.data.repository.BleDeviceRepository
import tech.mmarca.openvitals.devices.core.pairing.WatchPairingPort
import tech.mmarca.openvitals.domain.model.BleDeviceKind
import tech.mmarca.openvitals.domain.model.BleDiscoveredDevice
import tech.mmarca.openvitals.domain.model.BleSensorDevice
import tech.mmarca.openvitals.domain.model.DeviceIntegration

/** The one platform step WearOS onboarding shows. No bond, no probe. */
enum class WearOsOnboardStep { ASSOCIATING }

/** WearOS onboarding always succeeds. [associated] records the optional association. */
data class WearOsOnboardOutcome(
    val device: BleSensorDevice,
    val associated: Boolean,
)

/**
 * Turns a scanned WearOS watch into a registered `(WATCH, WEAROS)` device.
 * Associate (optional), then register with no capabilities, off the Garmin
 * sync path. Live heart rate comes over GATT, recorded data via Health Connect.
 */
@Singleton
class OnboardWearOsWatchUseCase @Inject constructor(
    private val pairing: WatchPairingPort,
    private val bleDeviceRepository: BleDeviceRepository,
) {

    /** [onStep] fires as the companion dialog is about to be shown. */
    suspend operator fun invoke(
        device: BleDiscoveredDevice,
        displayName: String,
        onStep: ((WearOsOnboardStep) -> Unit)? = null,
    ): WearOsOnboardOutcome {
        onStep?.invoke(WearOsOnboardStep.ASSOCIATING)
        val associated = try {
            pairing.associateCompanion(device.address, displayName)
        } catch (_: Exception) {
            // The association is best-effort; the watch onboards regardless.
            false
        }

        val registered = bleDeviceRepository.addDevice(
            displayName = displayName,
            address = device.address,
            bluetoothName = device.name,
            capabilities = emptySet(),
            kind = BleDeviceKind.WATCH,
            integration = DeviceIntegration.WEAROS,
        )
        return WearOsOnboardOutcome(device = registered, associated = associated)
    }

    /** Undoes [invoke] at the OS level: drops the association. */
    suspend fun forget(address: String) {
        pairing.disassociateCompanion(address)
    }
}
